src/smt/optimization_solver.h

```diff
@@ -122,32 +163,49 @@ class OptimizationSolver
    * Constructor
    * @param parent the smt_solver that the user added their assertions to
    **/
-  OptimizationSolver(SmtEngine* parent);
-  ~OptimizationSolver();
+  OptimizationSolver(SmtEngine* parent)
+      : d_parent(parent), d_objectives(), d_results()
+  {
+  }
+  ~OptimizationSolver() = default;
+
+  /** Runs the optimization loop for the pushed objective **/
+  OptimizationResult::ResultType checkOpt();
 
-  /** Runs the optimization loop for the activated objective **/
-  OptResult checkOpt();
   /**
-   * Activates an objective: will be optimized for
-   * @param obj the Node representing the expression that will be optimized for
+   * Pushes an objective: will be optimized for
+   * @param target the Node representing the expression that will be optimized
+   *for
    * @param type specifies whether it's maximize or minimize
    * @param bvSigned specifies whether we should use signed/unsigned
    *   comparison for BitVectors (only effective for BitVectors)
    *   and its default is false
    **/
-  void activateObj(const Node& obj,
-                   const ObjectiveType type,
-                   bool bvSigned = false);
-  /** Gets the value of the optimized objective after checkopt is called **/
-  Node objectiveGetValue();
+  void pushObjective(const Node target,
+                     const OptimizationObjective::ObjectiveType type,
+                     bool bvSigned = false);
+
+  /**
+   * Pops the objective that is lastly pushed.
+   **/
+  void popObjective();
+
+  /**
+   * Returns the values of the optimized objective after checkopt is called
+   * @return a vector of Optimization Result,
+   *   each containing the outcome and the value.
+   **/
+  std::vector<OptimizationResult> getValues();
 
  private:
   /** The parent SMT engine **/
   SmtEngine* d_parent;
+
   /** The objectives to optimize for **/
-  Objective d_activatedObjective;
-  /** A saved value of the objective from the last sat call. **/
-  Node d_savedValue;
+  std::vector<OptimizationObjective> d_objectives;
```

Comment:
Please clarify in the documentation of this field how are multiple objectives handled -- are they computed in the order they are presented in the vector? Are they independent of each other?

Reply:
Currently it only supports single obj.
and this will change as I push more PRs. 
there will be a parameter later specifying how multigoal is handled. But not this time. 

You know I am breaking a complete PR into chunks.

Reply:
detailed in comments